### ISSUE TITLE
Long-tailed datasets

### ISSUE THREAD

**Issue by PierreQuinton** [pytorch/vision#8093]
### 🚀 The feature

For each classification datasets with balanced distribution on the classes (MNIST, CIFAR-N, etc...), it would be very useful to provide a standard dataset for the imbalanced version of the dataset. For a dataset with $n$ classes, define the imbalance factor $a\in [0,1]$, then the proportion of class $i$ is typically be proportional to $a^{i/(n-1)}$, we need to normalize so that the proportions sums to $1$. For $a=1$ this is uniform and the smaller the imbalance coefficient the more imbalanced the dataset is.

I am not sure if torch vision should provide with the datasets or provide a data loader that imbalance the dataset.

### Motivation, pitch

Many papers are published on the problem of training on an imbalanced dataset and testing on a balanced dataset, for instance see [this](https://paperswithcode.com/task/long-tail-learning). As far as I know, there is no systematic way of generating such data sets for people using Pytorch. Here are few very similar implementations that are not fully satisfying :
- https://github.com/zhangyongshun/BagofTricks-LT/blob/main/lib/dataset/cao_cifar.py
- https://github.com/KaihuaTang/Long-Tailed-Recognition.pytorch/blob/master/classification/data/ImbalanceCIFAR.py

Such datasets seems to exist on TensorFlow, for instance section 3 of the readme of this [repo](https://github.com/google-research/google-research/tree/master/logit_adjustment) provides with links to download `tfrecord` datasets.

I feels like it could be a very nice feature of torchvision to either contain such datasets or be able to craft them easily.

**Comment by NicolasHug** [comment]
Hi @PierreQuinton ,

It seems like what you need is a [custom Sampler](https://pytorch.org/docs/stable/data.html#data-loading-order-and-sampler). IIUC, https://github.com/ufoym/imbalanced-dataset-sampler should be pretty close to what you're looking for?

**Comment by PierreQuinton** [comment]
@NicolasHug Thanks for your answer, yes this is exactly what I am looking for. I'm not sure if you would like to add something similar to torch or if you would close the issue, I leave it up to you.

**Comment by NicolasHug** [comment]
Thanks @PierreQuinton . I'll keep the issue open and rename it for clarity. Ultimately, what is needed to enable that is:

- 1) a consistent interface across datasets to access the idx -> classes mapping
- 2) a custom sampler (probably needed a distributed one as well)

1) is definitely in scope for torchvision and this is something we'd be doing if we ever re-start our work on a dataset revamp (CC @pmeier ).  For 2), we can decide when the time comes.

**Comment by justnotmegryan** [comment]
@PierreQuinton I have used synthetic data generation and UMAP to work out dataset imbalances quickly. Rendered.ai worked well for this.
